designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,218 @@
+	// +optional 
+	DisruptAfter *NillableDuration `json:"disruptAfter,omitempty"`
+	// Budgets is a list of Budgets. 
+	// If there are multiple active budgets, Karpenter uses 
+	// the most restrictive value. If left undefined, 
+	// this will default to one budget with a value to 10%. 
+	// +kubebuilder:validation:XValidation:message="'schedule' must be set with 'duration'",rule="!self.all(x, (has(x.schedule) && !has(x.duration)) || (!has(x.schedule) && has(x.duration)))" 
+	// +kubebuilder:default:={{nodes: "10%"}} 
+	// +kubebuilder:validation:MaxItems=50 
+	// +optional 
+	Budgets []Budget `json:"budgets,omitempty" hash:"ignore"` 
+}
+
+
+ // Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+	// Nodes dictates the maximum number of NodeClaims owned by this NodePool
+	// that can be terminating at once. This is calculated by counting nodes that
+	// have a deletion timestamp set, or are actively being deleted by Karpenter.
+	Nodes string `json:"nodes" hash:"ignore"`
+	// Schedule specifies when a budget begins being active, following
+	// the upstream cronjob syntax. If omitted, the budget is always active.
+	// Timezones are not supported.
+	Schedule *string `json:"schedule,omitempty" hash:"ignore"`
+	// Duration determines how long a Budget is active since each Schedule hit.
+	// Only minutes and hours are accepted, as cron does not work in seconds.
+	// If omitted, the budget is always active.
+	// This is required if Schedule is set.
+	Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+} // ... Same as Existing 
+
+type DisruptionAction string 
+
+const (
+	All DisruptionAction = "All"
+	Consolidation DisruptionAction = "Consolidation" 
+	Drift DisruptionAction = "Drift" 
+	Expiration DisruptionAction = "Expiration" 
+	Emptiness DisruptionAction = "Emptiness" ? 
+)
+```
+
+THis design proposes first defining Action Controls as a very barebones structure with a place in the api  
+```go
+// ActionControls defines the controls for a particular DisruptionAction, these controls are meant to apply for generic controls that apply to all disruption actions. 
+type ActionControls struct {
+	// Budgets is a list of Budgets. 
+	// If there are multiple active budgets, Karpenter uses 
+	// the most restrictive value. If left undefined, 
+	// this will default to one budget with a value to 10%. 
+	// +kubebuilder:validation:XValidation:message="'schedule' must be set with 'duration'",rule="!self.all(x, (has(x.schedule) && !has(x.duration)) || (!has(x.schedule) && has(x.duration)))" 
+	// +kubebuilder:default:={{nodes: "10%"}} 
+	// +kubebuilder:validation:MaxItems=50 
+	// +optional 
+	Budgets []Budget `json:"budgets,omitempty" hash:"ignore"` 
+}
+```
+
+#### Example 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    consolidationPolicy: WhenUnderutilized
```

Comment:
If we went with this design, it feels odd to not move the policy here into the `consolidation` block